File: TN-POS-SYS/TN-SYS-WEB/src/features/auth/services/s_api_auth_adm_role.ts
// 🇻🇳 API quản trị vai trò và phân quyền
// 🇺🇸 Role and permission management API
import { Api_Client } from '../../../core/api/api_client';
import type { M_Tb_Auth_Role } from '../data/models';

export class S_Api_Auth_Adm_Role {
  private apiClient: Api_Client;

  constructor(apiClient?: Api_Client) {
    this.apiClient = apiClient || new Api_Client();
  }

  // 🇻🇳 Lấy danh sách vai trò
  // 🇺🇸 Get role list
  async list(): Promise<M_Tb_Auth_Role[]> {
    const response = await this.apiClient.get<M_Tb_Auth_Role[]>('/auth/admin/roles');
    return response.data || [];
  }

  // 🇻🇳 Xóa vai trò
  // 🇺🇸 Delete role
  async delete(id: string): Promise<boolean> {
    const response = await this.apiClient.delete<boolean>(`/auth/admin/roles/${id}`);
    return response.data ?? false;
  }

  // 🇻🇳 Thêm/Sửa vai trò
  // 🇺🇸 Upsert role
  async upsert(role: M_Tb_Auth_Role): Promise<string> {
    const response = await this.apiClient.post<{ id: string }>('/auth/admin/roles', role);
    return response.data?.id || '';
  }

  // 🇻🇳 Lưu phân quyền cho vai trò
  // 🇺🇸 Save permissions for role
  async save_perms(roleId: string, permIds: string[]): Promise<boolean> {
    const response = await this.apiClient.post<boolean>('/auth/admin/roles/perms', {
      roleId,
      permIds,
    });
    return response.data ?? false;
  }
}

File: TN-POS-SYS/TN-SYS-FLU/lib/src/features/auth/services/s_api_auth_adm_role.dart
// 🇻🇳 API quản trị vai trò và phân quyền
// 🇺🇸 Role and permission management API
import '../../../core/api/api_client.dart';
import '../data/models/m_tb_auth_role.dart';

class S_Api_Auth_Adm_Role {
  final Api_Client apiClient;

  S_Api_Auth_Adm_Role({Api_Client? apiClient})
      : apiClient = apiClient ?? Api_Client();

  // 🇻🇳 Lấy danh sách vai trò
  // 🇺🇸 Get role list
  Future<List<M_Tb_Auth_Role>> list() async {
    final response = await apiClient.get<List<M_Tb_Auth_Role>>(
      '/auth/admin/roles',
      fromJson: (json) => (json as List)
          .map((e) => M_Tb_Auth_Role.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
    return response.data ?? [];
  }

  // 🇻🇳 Xóa vai trò
  // 🇺🇸 Delete role
  Future<bool> delete(String id) async {
    final response = await apiClient.delete<bool>(
      '/auth/admin/roles/$id',
      fromJson: (json) => json['data'] as bool? ?? false,
    );
    return response.data ?? false;
  }

  // 🇻🇳 Thêm/Sửa vai trò
  // 🇺🇸 Upsert role
  Future<String> upsert(M_Tb_Auth_Role role) async {
    final response = await apiClient.post<Map<String, dynamic>>(
      '/auth/admin/roles',
      body: role.toJson(),
      fromJson: (json) => json,
    );
    return response.data?['id'] as String? ?? '';
  }

  // 🇻🇳 Lưu phân quyền cho vai trò
  // 🇺🇸 Save permissions for role
  Future<bool> save_perms(String roleId, List<String> permIds) async {
    final response = await apiClient.post<bool>(
      '/auth/admin/roles/perms',
      body: {'roleId': roleId, 'permIds': permIds},
      fromJson: (json) => json['data'] as bool? ?? false,
    );
    return response.data ?? false;
  }
}

